### ISSUE TITLE
Rule blocked: CORERULES-4, CORERULES-48, CORERULES-49

### ISSUE THREAD

**Issue by chowsanthony** [cdisc-org/cdisc-rules-engine#152]
**Links to related JIRA Tickets**
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-4
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-49
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-48

**Rule Information**
- **Standard**: SDTMIG
- **Rule Description**:
  - CG0006: --DY calculated as per the study day algorithm as a non-zero integer value
  - CG0222: --ENDY is properly calculated per study day algorithm
  - CG0220: --STDY is properly calculated per study day algorithm

**Describe the problem**
The dy operator doesn't work for --DY variables as expected. For CG0006, it is flagging positive test case as issues. We realized the dy operator only handles --DY. We wanted it to also handle --STDY and --ENDY. Also, the new & improved dy operator needs to know how to handle incomplete or blank dates (see section below).

**Describe the solution**
Algorithms per cited guidance:
* From CG0006:
  * The Study Day value is incremented by 1 for each date following RFSTDTC. Dates prior to RFSTDTC are decreased by 1, with the date preceding RFSTDTC designated as Study Day -1 (there is no Study Day 0). . . . All Study Day values are integers. Thus, to calculate Study Day: --DY = (date portion of --DTC) - (date portion of RFSTDTC) + 1 if --DTC is on or after RFSTDTC --DY = (date portion of --DTC) - (date portion of RFSTDTC) if --DTC precedes RFSTDTC This algorithm should be used across all domains.
* From CG0220, CG0222:
  * The permissible Study Day variables (--DY, --STDY, and --ENDY) describe the relative day of the observation starting with the reference date as Day 1. They are determined by comparing the date portion of the respective date/time variables (--DTC, --STDTC, and --ENDTC) to the date portion of the Subject Reference Start Date (RFSTDTC from the Demographics domain).][All Study Day values are integers. Thus, to calculate Study Day: --DY = (date portion of --DTC) - (date portion of RFSTDTC) + 1 if --DTC is on or after RFSTDTC, --DY = (date portion of --DTC) - (date portion of RFSTDTC) if --DTC precedes RFSTDTC. This algorithm should be used across all domains.
* When either date (RFSTDTC or {--DTC | --STDTC | --ENDTC}) is null or incomplete, the resulting DY variable is expected to be null.

**Proposed rule logic**
```yaml
Check:
  all:
    - name: --DY
      operator: not_equal_to
      value: $val_dy
Operations:
  - name: RFSTDTC
    domain: DM
    operator: dy
    id: $val_dy
Scopes:
  Classes:
    Include:
      - All
  Domains:
    Include:
      - All
  Standards:
    - Name: SDTMIG
      Version: "3.4"
```

```yaml
Check:
  all:
    - name: --STDY
      operator: not_equal_to
      value: $val_stdy
Operations:
  - name: RFSTDTC
    domain: DM
    operator: dy
    id: $val_stdy
Scopes:
  Classes:
    Include:
      - All
  Domains:
    Exclude:
      - DM
  Standards:
    - Name: SDTMIG
      Version: "3.4"
```

```yaml
Check:
  all:
    - name: --ENDY
      operator: not_equal_to
      value: $val_endy
Operations:
  - name: RFSTDTC
    domain: DM
    operator: dy
    id: $val_endy
Scopes:
  Classes:
    Include:
      - All
  Domains:
    Exclude:
      - DM
  Standards:
    - Name: SDTMIG
      Version: "3.4"
```

**Screenshots**
If applicable, add screenshots to help explain your problem and solution.
[dy-related-rules.zip](https://github.com/cdisc-org/cdisc-rules-engine/files/9709093/dy-related-rules.zip)

**Comment by gcampion-cdisc** [comment]
After investigating some more, I think the issue here is a bit different from what is described. 2 things:
1. the rule is written incorrectly - it should have `--STDTC` in the `name` instead of `RFSTDTC` for the `dy` operation
2. There needs to be a way to exclude `domain` from the `dy` operation. When domain is excluded, the domain should be the current domain. The dy operation should continue to assume we are anchoring with RFSTDTC, which is found in DM.

For more information, you can refer to the discussion here: https://dev.azure.com/cdisc-org/cdisc-core/_workitems/edit/2538
I guess a followup ticket was never created.
